src/lib/QtWidgets/QTreeWidgetItem.ts

```diff
@@ -0,0 +1,38 @@
+import addon from '../utils/addon';
+import { Component, NativeElement } from '../core/Component';
+import { QTreeWidget } from '../..';
+
+export class QTreeWidgetItem extends Component {
+    native: NativeElement;
+
+    constructor();
+    constructor(strings?: string[]); // 1
```

Comment:
Another point that I am wondering is that it is more preferable to have lean constructor and have more methods that do the same operation since this will be a requirement when implementing the same widget in the react nodegui world, where on prop change we will need to perform operations and not create an new instance again.

Reply:
I agree, will make one constructor, which will be QWidgetTree or QWidgetTreeItem with optional text, seems much more cleaner.

Reply:
```
    constructor();
    constructor(parent?: QTreeWidgetItem | QTreeWidget, strings?: string[]) {
        super();
        this.items = new Set();
        if (parent) {
            this.nodeParent = parent;
            if (strings) {
                this.native = new addon.QTreeWidgetItem(parent.native, strings);
            } else {
                this.native = new addon.QTreeWidgetItem(parent.native);
            }
            parent.items.add(this);
        } else if (strings) {
            this.native = new addon.QTreeWidgetItem(strings);
        } else {
            this.native = new addon.QTreeWidgetItem();
        }
    }

```

I think this looks much better, what do you think?

Reply:
Doing it like this, how would I unwrap the parent object in C++ being that it could be either QTreeWidgetItem, or QTreeWidget?

Reply:
```
QTreeWidgetItemWrap::QTreeWidgetItemWrap(const Napi::CallbackInfo& info): Napi::ObjectWrap<QTreeWidgetItemWrap>(info) {
    Napi::Env env = info.Env();
    Napi::HandleScope scope(env);

    if (info.Length() == 2) {
        Napi::Array stringsNapi = info[1].As<Napi::Array>();
        QList<QString> list;
        for (int i = 0; i < stringsNapi.Length(); i++) {
        Napi::Value stringNapi = stringsNapi[i];
            list.append(stringNapi.As<Napi::String>().Utf8Value().c_str());
        }
        QStringList strings = QStringList(list);

        try{
            Napi::Object parentObject = info[0].As<Napi::Object>();
            QTreeWidgetWrap* parentWidgetWrap = Napi::ObjectWrap<QTreeWidgetWrap>::Unwrap(parentObject);
            QTreeWidget* parent = parentWidgetWrap->getInternalInstance();
            this->instance = new QTreeWidgetItem(parent);
        } catch (int n) {
          Napi::Object itemObject = info[0].As<Napi::Object>();
          QTreeWidgetItemWrap* itemWidgetWrap = Napi::ObjectWrap<QTreeWidgetItemWrap>::Unwrap(itemObject);
          QTreeWidgetItem* item = itemWidgetWrap->getInternalInstance();
          this->instance = new QTreeWidgetItem(item);
        }
    } else if(info.Length() == 1){
        if (info[0].IsArray()){
            Napi::Array stringsNapi = info[1].As<Napi::Array>();
            QList<QString> list;
            for (int i = 0; i < stringsNapi.Length(); i++) {
                Napi::Value stringNapi = stringsNapi[i];
                list.append(stringNapi.As<Napi::String>().Utf8Value().c_str());
                }
            QStringList strings = QStringList(list);
            this->instance = new QTreeWidgetItem(strings);
        } else {
            try{
                Napi::Object parentObject = info[0].As<Napi::Object>();
                QTreeWidgetWrap* parentWidgetWrap = Napi::ObjectWrap<QTreeWidgetWrap>::Unwrap(parentObject);
                QTreeWidget* parent = parentWidgetWrap->getInternalInstance();
                this->instance = new QTreeWidgetItem(parent);
            } catch (int n) {
              Napi::Object itemObject = info[0].As<Napi::Object>();
              QTreeWidgetItemWrap* itemWidgetWrap = Napi::ObjectWrap<QTreeWidgetItemWrap>::Unwrap(itemObject);
              QTreeWidgetItem* item = itemWidgetWrap->getInternalInstance();
              this->instance = new QTreeWidgetItem(item);
            }
        }
    } else if (info.Length() == 0) {
        this->instance = new QTreeWidgetItem();
    } else {
        Napi::TypeError::New(env, "Wrong number of arguments").ThrowAsJavaScriptException();
    }
    this->rawData = extrautils::configureComponent(this->getInternalInstance());
}

```

I currently have it like this but there has to be a cleaner way to do it. @master-atul

Reply:
oh, let send in a string or an enum from the js side after testing it whether  its an instance of item or widget instead of relying on the try catch. its not perfect but definetly better. @mspencer92

Reply:
if its too much work you can also leave it as it is and I can take a look at it when I get some time aswell

Reply:
> oh, let send in a string or an enum from the js side after testing it whether its an instance of item or widget instead of relying on the try catch. its not perfect but definetly better. @mspencer92

That is exactly what I did, the try except wasn't working correctly. The new push is working great. I need to start working on functions and signals. Thank you for all the help @master-atul

Reply:
> oh, let send in a string or an enum from the js side after testing it whether its an instance of item or widget instead of relying on the try catch. its not perfect but definetly better. @mspencer92

That is exactly what I did, the try except wasn't working correctly. The new push is working great. I need to start working on functions and signals. Thank you for all the help @master-atul.

The constructor is now as ` constructor(parent?: QTreeWidgetItem | QTreeWidget, strings?: string[]) {` 
How could I change it so that parent could be empty and QTreeWidgetItem can be initialized as QTreeWidgetItem(['string', 'list']);

Reply:
I did not quite get this @mspencer92?
Isn't your contructor for QTreeWidget already ready for all these cases? Can you explain it a bit more?

Reply:
> 
> 
> I did not quite get this @mspencer92?
> Isn't your contructor for QTreeWidget already ready for all these cases? Can you explain it a bit more?

I want to be able to do QTreeWidgetItem(["Text"]); without specifying a parent, and the current constructor that I have does not support that.

Reply:
@master-atul is there another way I can contact you besides here? This seems a little slow and I don't always get your replies when you send them.

Reply:
Oh yes, we can chat on spectrum maybe ? or twitter ?

Reply:
I have no idea what spectrum is other than an internet and cable provider.

Reply:
May

> > I did not quite get this @mspencer92?
> > Isn't your contructor for QTreeWidget already ready for all these cases? Can you explain it a bit more?
> 
> I want to be able to do QTreeWidgetItem(["Text"]); without specifying a parent, and the current constructor that I have does not support that.
> 
> I would like to be able to do
> const item = new QTreeWidgetItem(['Hello!']);
> tree.addTopLevelItem(item);

Maybe check if its an array 
```
 } else if (strings) { <----- here
            this.native = new addon.QTreeWidgetItem(strings);
        } else {
```

Reply:
Got it so change the constructor to
` constructor(arg?: QTreeWidgetItem | QTreeWidget | string[], strings?: string[]) {` and then check if arg is an array.

Reply:
Yep 👍

Reply:
Thanks, that works good. I think I am happy with the constructor now to move onto functions and signals.